server_test.go

```diff
@@ -4,41 +4,155 @@ import (
 	"encoding/json"
 	"io"
 	"net"
+
+	"github.com/hashicorp/vault/audit"
 )
 
-func setupHander(ts *TestSuite) (net.Conn, *AuditEntryQueue) {
-	server, client := net.Pipe()
-	ts.AddCleanup(func() { ts.Nil(server.Close()) })
+type auditEntryCollector struct {
+	requests  []*audit.AuditRequestEntry
+	responses []*audit.AuditResponseEntry
+}
 
-	queue := NewAuditEntryQueue()
-	ts.AddCleanup(queue.Close)
+func newAuditEntryCollector() *auditEntryCollector {
+	return &auditEntryCollector{
+		requests:  make([]*audit.AuditRequestEntry, 0),
+		responses: make([]*audit.AuditResponseEntry, 0),
+	}
+}
 
-	go handleConnection(client, queue)
+func (col *auditEntryCollector) HandleRequest(req *audit.AuditRequestEntry) {
+	col.requests = append(col.requests, req)
+}
+
+func (col *auditEntryCollector) HandleResponse(res *audit.AuditResponseEntry) {
+	col.responses = append(col.responses, res)
+}
 
-	return server, queue
+func writeJSONLine(v interface{}, writer io.Writer) {
+	b, _ := json.Marshal(v)
+	writeLine(b, writer)
 }
 
-func writeJSONLine(ts *TestSuite, v interface{}, writer io.Writer) {
-	b, _ := ts.WithoutError(json.Marshal(v)).([]byte)
+func writeStringLine(s string, writer io.Writer) {
+	writeLine([]byte(s), writer)
+}
 
-	ts.WithoutError(writer.Write(b))
-	ts.WithoutError(writer.Write([]byte("\n")))
+func writeLine(b []byte, writer io.Writer) {
+	_, _ = writer.Write(b)
+	_, _ = writer.Write([]byte{'\n'})
 }
 
 func (ts *TestSuite) TestHandleRequest() {
-	server, queue := setupHander(ts)
+	server, client := net.Pipe()
 
 	req := dummyRequest()
-	writeJSONLine(ts, req, server)
+	go func() {
+		defer server.Close()
+		writeJSONLine(req, server)
+	}()
 
-	ts.Equal(req, <-queue.Receive())
+	collector := newAuditEntryCollector()
+	handleConnection(client, collector)
+
+	ts.Len(collector.requests, 1)
+	ts.Equal(req, collector.requests[0])
 }
 
 func (ts *TestSuite) TestHandleResponse() {
-	server, queue := setupHander(ts)
+	server, client := net.Pipe()
+
+	res := dummyResponse()
+	go func() {
+		defer server.Close()
+		writeJSONLine(res, server)
+	}()
+
+	collector := newAuditEntryCollector()
+	handleConnection(client, collector)
+
+	ts.Len(collector.responses, 1)
+	ts.Equal(res, collector.responses[0])
+}
+
+func (ts *TestSuite) TestUnknownTypeIgnored() {
+	server, client := net.Pipe()
 
+	req := dummyRequest()
 	res := dummyResponse()
-	writeJSONLine(ts, res, server)
+	go func() {
+		defer server.Close()
+		writeJSONLine(req, server)
+		writeStringLine("{\"type\": \"foo\"}", server)
+		writeJSONLine(res, server)
+	}()
+
+	collector := newAuditEntryCollector()
+	handleConnection(client, collector)
 
+	ts.Len(collector.requests, 1)
+	ts.Equal(req, collector.requests[0])
+	ts.Len(collector.responses, 1)
+	ts.Equal(res, collector.responses[0])
+}
+
+func (ts *TestSuite) TestUnknownJSON() {
+	server, client := net.Pipe()
+	defer server.Close()
+
+	go func() {
+		writeStringLine("{\"foo\": \"bar\"}", server)
```

Comment:
Do we maybe want to surround this with known-good events to make sure the bad one doesn't break us? (Similarly for other tests where it's a possibility, or maybe just one big test in which we send all kinds of badness interspersed with good events that we can track as markers.)

Reply:
The bad ones do "break" us. The `handleConnection` function returns and the connection is closed. The only case where we ignore/log an error and continue is if the `type` is something unknown, in case Vault adds a new type of audit entry or something.

Reply:
This is maybe my fault for not writing enough documentation.

Reply:
Maybe add a valid entry to the end of the list to assert that these kinds of bad entries break us? That'll confirm that the behaviour is what we expect either way.

Reply:
The write should fail, though, since the connection is closed. But, yes, I guess I should assert that.